File: go/ontology/store_test.go
// SPDX-License-Identifier: Apache-2.0
package ontology_test

import (
	"context"
	"encoding/json"
	"testing"

	"github.com/jeffs-brain/memory/go/brain"
	"github.com/jeffs-brain/memory/go/ontology"
	"github.com/jeffs-brain/memory/go/store/mem"
)

func newTestStore(t *testing.T) (*ontology.FileOntologyStore, *mem.Store) {
	t.Helper()
	ms := mem.New()
	t.Cleanup(func() { _ = ms.Close() })
	cfg := ontology.FileOntologyStoreConfig{
		BrainID:   "brain-1",
		ProjectID: "proj-1",
		OrgID:     "org-1",
	}
	store, err := ontology.NewFileOntologyStore(ms, cfg)
	if err != nil {
		t.Fatalf("NewFileOntologyStore: %v", err)
	}
	return store, ms
}

func TestGetType_BuiltInNodeType(t *testing.T) {
	t.Parallel()
	s, _ := newTestStore(t)
	ctx := context.Background()

	def, err := s.GetType(ctx, ontology.ScopeBuiltIn, "entity.customer")
	if err != nil {
		t.Fatalf("GetType(built-in, entity.customer) = %v", err)
	}
	if def.Type != "entity.customer" {
		t.Fatalf("expected type entity.customer, got %q", def.Type)
	}
	if def.Status != ontology.TypeStatusActive {
		t.Fatalf("expected status active, got %q", def.Status)
	}
	if def.Label == "" {
		t.Fatal("expected non-empty label")
	}
	if def.Description == "" {
		t.Fatal("expected non-empty description")
	}
}

func TestGetType_BuiltInEdgeType(t *testing.T) {
	t.Parallel()
	s, _ := newTestStore(t)
	ctx := context.Background()

	def, err := s.GetType(ctx, ontology.ScopeBuiltIn, "triggers")
	if err != nil {
		t.Fatalf("GetType(built-in, triggers) = %v", err)
	}
	if def.Type != "triggers" {
		t.Fatalf("expected type triggers, got %q", def.Type)
	}
}

func TestGetType_BuiltInNotFound(t *testing.T) {
	t.Parallel()
	s, _ := newTestStore(t)
	ctx := context.Background()

	_, err := s.GetType(ctx, ontology.ScopeBuiltIn, "nonexistent.type")
	if err == nil {
		t.Fatal("expected error for nonexistent built-in type")
	}
}

func TestUpsertType_BrainScope(t *testing.T) {
	t.Parallel()
	s, _ := newTestStore(t)
	ctx := context.Background()

	def := ontology.TypeDefinition{
		Type:        "entity.invoice",
		Label:       "Invoice",
		Description: "Financial invoice documents",
		CreatedAt:   "2026-05-10T00:00:00.000Z",
		Status:      ontology.TypeStatusActive,
	}
	if err := s.UpsertType(ctx, ontology.ScopeBrain, def); err != nil {
		t.Fatalf("UpsertType = %v", err)
	}

	got, err := s.GetType(ctx, ontology.ScopeBrain, "entity.invoice")
	if err != nil {
		t.Fatalf("GetType after upsert = %v", err)
	}
	if got.Type != "entity.invoice" {
		t.Fatalf("expected entity.invoice, got %q", got.Type)
	}
	if got.Label != "Invoice" {
		t.Fatalf("expected label Invoice, got %q", got.Label)
	}
}

func TestUpsertType_OverwritesExisting(t *testing.T) {
	t.Parallel()
	s, _ := newTestStore(t)
	ctx := context.Background()

	def := ontology.TypeDefinition{
		Type:        "entity.invoice",
		Label:       "Invoice V1",
		Description: "First version",
		CreatedAt:   "2026-05-10T00:00:00.000Z",
		Status:      ontology.TypeStatusActive,
	}
	if err := s.UpsertType(ctx, ontology.ScopeBrain, def); err != nil {
		t.Fatalf("first UpsertType = %v", err)
	}

	def.Label = "Invoice V2"
	def.Description = "Second version"
	if err := s.UpsertType(ctx, ontology.ScopeBrain, def); err != nil {
		t.Fatalf("second UpsertType = %v", err)
	}

	got, err := s.GetType(ctx, ontology.ScopeBrain, "entity.invoice")
	if err != nil {
		t.Fatalf("GetType after overwrite = %v", err)
	}
	if got.Label != "Invoice V2" {
		t.Fatalf("expected Invoice V2, got %q", got.Label)
	}
}

func TestUpsertType_RejectsBuiltInScope(t *testing.T) {
	t.Parallel()
	s, _ := newTestStore(t)
	ctx := context.Background()

	def := ontology.TypeDefinition{
		Type:        "entity.test",
		Label:       "Test",
		Description: "A test type",
		CreatedAt:   "2026-05-10T00:00:00.000Z",
		Status:      ontology.TypeStatusActive,
	}
	err := s.UpsertType(ctx, ontology.ScopeBuiltIn, def)
	if err == nil {
		t.Fatal("expected error when upserting to built-in scope")
	}
}

func TestUpsertType_RejectsInvalidDefinition(t *testing.T) {
	t.Parallel()
	s, _ := newTestStore(t)
	ctx := context.Background()

	def := ontology.TypeDefinition{
		Type:   "entity.test",
		Label:  "",
		Status: ontology.TypeStatusActive,
	}
	err := s.UpsertType(ctx, ontology.ScopeBrain, def)
	if err == nil {
		t.Fatal("expected error for invalid definition")
	}
}

func TestDeleteType_RemovesFromScope(t *testing.T) {
	t.Parallel()
	s, _ := newTestStore(t)
	ctx := context.Background()

	def := ontology.TypeDefinition{
		Type:        "entity.invoice",
		Label:       "Invoice",
		Description: "Financial invoices",
		CreatedAt:   "2026-05-10T00:00:00.000Z",
		Status:      ontology.TypeStatusActive,
	}
	if err := s.UpsertType(ctx, ontology.ScopeBrain, def); err != nil {
		t.Fatalf("UpsertType = %v", err)
	}

	if err := s.DeleteType(ctx, ontology.ScopeBrain, "entity.invoice"); err != nil {
		t.Fatalf("DeleteType = %v", err)
	}

	_, err := s.GetType(ctx, ontology.ScopeBrain, "entity.invoice")
	if err == nil {
		t.Fatal("expected error after delete")
	}
}

func TestDeleteType_NotFoundReturnsError(t *testing.T) {
	t.Parallel()
	s, _ := newTestStore(t)
	ctx := context.Background()

	err := s.DeleteType(ctx, ontology.ScopeBrain, "entity.nonexistent")
	if err == nil {
		t.Fatal("expected error when deleting nonexistent type")
	}
}

func TestDeleteType_RejectsBuiltInScope(t *testing.T) {
	t.Parallel()
	s, _ := newTestStore(t)
	ctx := context.Background()

	err := s.DeleteType(ctx, ontology.ScopeBuiltIn, "entity.customer")
	if err == nil {
		t.Fatal("expected error when deleting from built-in scope")
	}
}

func TestListTypes_BuiltIn(t *testing.T) {
	t.Parallel()
	s, _ := newTestStore(t)
	ctx := context.Background()

	types, err := s.ListTypes(ctx, ontology.ScopeBuiltIn, ontology.ListTypesOpts{})
	if err != nil {
		t.Fatalf("ListTypes(built-in) = %v", err)
	}
	if len(types) != 50 {
		t.Fatalf("expected 50 built-in types (31 nodes + 19 edges), got %d", len(types))
	}
}

func TestListTypes_FilterByPrefix(t *testing.T) {
	t.Parallel()
	s, _ := newTestStore(t)
	ctx := context.Background()

	types, err := s.ListTypes(ctx, ontology.ScopeBuiltIn, ontology.ListTypesOpts{Prefix: "entity."})
	if err != nil {
		t.Fatalf("ListTypes(prefix=entity.) = %v", err)
	}
	if len(types) != 6 {
		t.Fatalf("expected 6 entity types, got %d", len(types))
	}
	for _, td := range types {
		if !hasEntityPrefix(td.Type) {
			t.Fatalf("unexpected type %q in entity prefix filter", td.Type)
		}
	}
}

func TestListTypes_FilterByStatus(t *testing.T) {
	t.Parallel()
	s, _ := newTestStore(t)
	ctx := context.Background()

	active := ontology.TypeDefinition{
		Type:        "entity.active_one",
		Label:       "Active One",
		Description: "An active type",
		CreatedAt:   "2026-05-10T00:00:00.000Z",
		Status:      ontology.TypeStatusActive,
	}
	proposed := ontology.TypeDefinition{
		Type:        "entity.proposed_one",
		Label:       "Proposed One",
		Description: "A proposed type",
		CreatedAt:   "2026-05-10T00:00:00.000Z",
		Status:      ontology.TypeStatusProposed,
	}
	if err := s.UpsertType(ctx, ontology.ScopeBrain, active); err != nil {
		t.Fatalf("UpsertType active = %v", err)
	}
	if err := s.UpsertType(ctx, ontology.ScopeBrain, proposed); err != nil {
		t.Fatalf("UpsertType proposed = %v", err)
	}

	types, err := s.ListTypes(ctx, ontology.ScopeBrain, ontology.ListTypesOpts{Status: "proposed"})
	if err != nil {
		t.Fatalf("ListTypes(status=proposed) = %v", err)
	}
	if len(types) != 1 {
		t.Fatalf("expected 1 proposed type, got %d", len(types))
	}
	if types[0].Type != "entity.proposed_one" {
		t.Fatalf("expected entity.proposed_one, got %q", types[0].Type)
	}
}

func TestGetResolvedOntology_BuiltInOnly(t *testing.T) {
	t.Parallel()
	s, _ := newTestStore(t)
	ctx := context.Background()

	resolved, err := s.GetResolvedOntology(ctx, "brain-1", "proj-1", "org-1")
	if err != nil {
		t.Fatalf("GetResolvedOntology = %v", err)
	}
	if len(resolved.NodeTypes) != 31 {
		t.Fatalf("expected 31 node types, got %d", len(resolved.NodeTypes))
	}
	if len(resolved.EdgeTypes) != 19 {
		t.Fatalf("expected 19 edge types, got %d", len(resolved.EdgeTypes))
	}
	if len(resolved.BusinessCategories) != 8 {
		t.Fatalf("expected 8 categories, got %d", len(resolved.BusinessCategories))
	}
}

func TestGetResolvedOntology_BrainOverridesBuiltIn(t *testing.T) {
	t.Parallel()
	s, _ := newTestStore(t)
	ctx := context.Background()

	override := ontology.TypeDefinition{
		Type:        "entity.customer",
		Label:       "Custom Customer",
		Description: "Brain-level override of built-in customer type",
		CreatedAt:   "2026-05-10T00:00:00.000Z",
		Status:      ontology.TypeStatusActive,
	}
	if err := s.UpsertType(ctx, ontology.ScopeBrain, override); err != nil {
		t.Fatalf("UpsertType = %v", err)
	}

	resolved, err := s.GetResolvedOntology(ctx, "brain-1", "proj-1", "org-1")
	if err != nil {
		t.Fatalf("GetResolvedOntology = %v", err)
	}

	for _, rt := range resolved.NodeTypes {
		if rt.Type == "entity.customer" {
			if rt.Label != "Custom Customer" {
				t.Fatalf("expected brain override label, got %q", rt.Label)
			}
			if rt.Scope != ontology.ScopeBrain {
				t.Fatalf("expected scope brain, got %q", rt.Scope)
			}
			return
		}
	}
	t.Fatal("entity.customer not found in resolved ontology")
}

func TestGetResolvedOntology_ScopeHierarchy(t *testing.T) {
	t.Parallel()
	s, _ := newTestStore(t)
	ctx := context.Background()

	orgDef := ontology.TypeDefinition{
		Type:        "entity.tenant",
		Label:       "Org Tenant",
		Description: "Organisation-level tenant",
		CreatedAt:   "2026-05-10T00:00:00.000Z",
		Status:      ontology.TypeStatusActive,
	}
	if err := s.UpsertType(ctx, ontology.ScopeOrganisation, orgDef); err != nil {
		t.Fatalf("UpsertType org = %v", err)
	}

	projectDef := ontology.TypeDefinition{
		Type:        "entity.tenant",
		Label:       "Project Tenant",
		Description: "Project-level tenant override",
		CreatedAt:   "2026-05-10T00:00:00.000Z",
		Status:      ontology.TypeStatusActive,
	}
	if err := s.UpsertType(ctx, ontology.ScopeProject, projectDef); err != nil {
		t.Fatalf("UpsertType project = %v", err)
	}

	brainDef := ontology.TypeDefinition{
		Type:        "entity.tenant",
		Label:       "Brain Tenant",
		Description: "Brain-level tenant override",
		CreatedAt:   "2026-05-10T00:00:00.000Z",
		Status:      ontology.TypeStatusActive,
	}
	if err := s.UpsertType(ctx, ontology.ScopeBrain, brainDef); err != nil {
		t.Fatalf("UpsertType brain = %v", err)
	}

	resolved, err := s.GetResolvedOntology(ctx, "brain-1", "proj-1", "org-1")
	if err != nil {
		t.Fatalf("GetResolvedOntology = %v", err)
	}

	for _, rt := range resolved.NodeTypes {
		if rt.Type == "entity.tenant" {
			if rt.Label != "Brain Tenant" {
				t.Fatalf("expected brain to shadow project and org, got label %q", rt.Label)
			}
			if rt.Scope != ontology.ScopeBrain {
				t.Fatalf("expected scope brain, got %q", rt.Scope)
			}
			return
		}
	}
	t.Fatal("entity.tenant not found in resolved ontology")
}

func TestGetResolvedOntology_ProjectOverridesOrg(t *testing.T) {
	t.Parallel()
	s, _ := newTestStore(t)
	ctx := context.Background()

	orgDef := ontology.TypeDefinition{
		Type:        "entity.account",
		Label:       "Org Account",
		Description: "Organisation-level account",
		CreatedAt:   "2026-05-10T00:00:00.000Z",
		Status:      ontology.TypeStatusActive,
	}
	if err := s.UpsertType(ctx, ontology.ScopeOrganisation, orgDef); err != nil {
		t.Fatalf("UpsertType org = %v", err)
	}

	projectDef := ontology.TypeDefinition{
		Type:        "entity.account",
		Label:       "Project Account",
		Description: "Project-level account override",
		CreatedAt:   "2026-05-10T00:00:00.000Z",
		Status:      ontology.TypeStatusActive,
	}
	if err := s.UpsertType(ctx, ontology.ScopeProject, projectDef); err != nil {
		t.Fatalf("UpsertType project = %v", err)
	}

	resolved, err := s.GetResolvedOntology(ctx, "brain-1", "proj-1", "org-1")
	if err != nil {
		t.Fatalf("GetResolvedOntology = %v", err)
	}

	for _, rt := range resolved.NodeTypes {
		if rt.Type == "entity.account" {
			if rt.Label != "Project Account" {
				t.Fatalf("expected project to shadow org, got label %q", rt.Label)
			}
			if rt.Scope != ontology.ScopeProject {
				t.Fatalf("expected scope project, got %q", rt.Scope)
			}
			return
		}
	}
	t.Fatal("entity.account not found in resolved ontology")
}

func TestGetResolvedOntology_DeprecatedExcluded(t *testing.T) {
	t.Parallel()
	s, _ := newTestStore(t)
	ctx := context.Background()

	deprecated := ontology.TypeDefinition{
		Type:        "entity.legacy",
		Label:       "Legacy Entity",
		Description: "A deprecated type",
		CreatedAt:   "2026-05-10T00:00:00.000Z",
		Status:      ontology.TypeStatusDeprecated,
	}
	if err := s.UpsertType(ctx, ontology.ScopeBrain, deprecated); err != nil {
		t.Fatalf("UpsertType = %v", err)
	}

	resolved, err := s.GetResolvedOntology(ctx, "brain-1", "proj-1", "org-1")
	if err != nil {
		t.Fatalf("GetResolvedOntology = %v", err)
	}

	for _, rt := range resolved.NodeTypes {
		if rt.Type == "entity.legacy" {
			t.Fatal("deprecated type should not appear in resolved ontology")
		}
	}
}

func TestGetResolvedOntology_CustomCategories(t *testing.T) {
	t.Parallel()
	_, ms := newTestStore(t)
	cfg := ontology.FileOntologyStoreConfig{
		BrainID:   "brain-1",
		ProjectID: "proj-1",
		OrgID:     "org-1",
	}
	s, err := ontology.NewFileOntologyStore(ms, cfg)
	if err != nil {
		t.Fatalf("NewFileOntologyStore: %v", err)
	}
	ctx := context.Background()

	stored := ontology.StoredOntology{
		CustomBusinessCategories: []string{"logistics", "healthcare"},
	}
	data, err := json.Marshal(stored)
	if err != nil {
		t.Fatalf("marshal = %v", err)
	}
	if err := ms.Write(ctx, brain.Path("ontology/org/org-1/types.json"), data); err != nil {
		t.Fatalf("Write = %v", err)
	}

	resolved, err := s.GetResolvedOntology(ctx, "brain-1", "proj-1", "org-1")
	if err != nil {
		t.Fatalf("GetResolvedOntology = %v", err)
	}

	categorySet := make(map[string]struct{}, len(resolved.BusinessCategories))
	for _, c := range resolved.BusinessCategories {
		categorySet[c] = struct{}{}
	}
	if _, ok := categorySet["logistics"]; !ok {
		t.Fatal("expected logistics category in resolved ontology")
	}
	if _, ok := categorySet["healthcare"]; !ok {
		t.Fatal("expected healthcare category in resolved ontology")
	}
	if _, ok := categorySet["customer"]; !ok {
		t.Fatal("expected built-in customer category in resolved ontology")
	}
}

func TestGetResolvedOntology_Cached(t *testing.T) {
	t.Parallel()
	s, _ := newTestStore(t)
	ctx := context.Background()

	resolved1, err := s.GetResolvedOntology(ctx, "brain-1", "proj-1", "org-1")
	if err != nil {
		t.Fatalf("first GetResolvedOntology = %v", err)
	}
	resolved2, err := s.GetResolvedOntology(ctx, "brain-1", "proj-1", "org-1")
	if err != nil {
		t.Fatalf("second GetResolvedOntology = %v", err)
	}
	if resolved1 != resolved2 {
		t.Fatal("expected cached pointer equality")
	}
}

func TestGetResolvedOntology_CacheInvalidatedOnUpsert(t *testing.T) {
	t.Parallel()
	s, _ := newTestStore(t)
	ctx := context.Background()

	resolved1, err := s.GetResolvedOntology(ctx, "brain-1", "proj-1", "org-1")
	if err != nil {
		t.Fatalf("first GetResolvedOntology = %v", err)
	}

	def := ontology.TypeDefinition{
		Type:        "entity.fresh",
		Label:       "Fresh",
		Description: "A fresh type",
		CreatedAt:   "2026-05-10T00:00:00.000Z",
		Status:      ontology.TypeStatusActive,
	}
	if err := s.UpsertType(ctx, ontology.ScopeBrain, def); err != nil {
		t.Fatalf("UpsertType = %v", err)
	}

	resolved2, err := s.GetResolvedOntology(ctx, "brain-1", "proj-1", "org-1")
	if err != nil {
		t.Fatalf("second GetResolvedOntology = %v", err)
	}
	if resolved1 == resolved2 {
		t.Fatal("expected cache to be invalidated after upsert")
	}
}

func TestUpsertType_EdgeType(t *testing.T) {
	t.Parallel()
	s, _ := newTestStore(t)
	ctx := context.Background()

	def := ontology.TypeDefinition{
		Type:        "manages",
		Label:       "Manages",
		Description: "Source manages target",
		CreatedAt:   "2026-05-10T00:00:00.000Z",
		Status:      ontology.TypeStatusActive,
	}
	if err := s.UpsertType(ctx, ontology.ScopeBrain, def); err != nil {
		t.Fatalf("UpsertType edge = %v", err)
	}

	got, err := s.GetType(ctx, ontology.ScopeBrain, "manages")
	if err != nil {
		t.Fatalf("GetType edge = %v", err)
	}
	if got.Type != "manages" {
		t.Fatalf("expected manages, got %q", got.Type)
	}
}

func hasEntityPrefix(s string) bool {
	return len(s) > 7 && s[:7] == "entity."
}

File: sdks/ts/memory/src/ontology/store.test.ts
// SPDX-License-Identifier: Apache-2.0

import { beforeEach, describe, expect, it } from 'vitest'
import { createMemStore } from '../store/memstore.js'
import type { Store } from '../store/index.js'
import { toPath } from '../store/path.js'
import type { OntologyTypeDefinition } from './types.js'
import {
  createFileOntologyStore,
  type FileOntologyStoreConfig,
  type OntologyStore,
} from './store.js'

const TEST_CONFIG: FileOntologyStoreConfig = {
  brainId: 'brain-1',
  projectId: 'proj-1',
  orgId: 'org-1',
}

function makeStore(): { ontologyStore: OntologyStore; backingStore: Store } {
  const backingStore = createMemStore()
  const ontologyStore = createFileOntologyStore(backingStore, TEST_CONFIG)
  return { ontologyStore, backingStore }
}

function makeNodeDef(overrides: Partial<OntologyTypeDefinition> = {}): OntologyTypeDefinition {
  return {
    type: 'entity.invoice',
    label: 'Invoice',
    description: 'Financial invoice documents',
    createdAt: '2026-05-10T00:00:00.000Z',
    status: 'active',
    ...overrides,
  }
}

function makeEdgeDef(overrides: Partial<OntologyTypeDefinition> = {}): OntologyTypeDefinition {
  return {
    type: 'manages',
    label: 'Manages',
    description: 'Source manages target',
    createdAt: '2026-05-10T00:00:00.000Z',
    status: 'active',
    ...overrides,
  }
}

describe('FileOntologyStore', () => {
  let ontologyStore: OntologyStore
  let backingStore: Store

  beforeEach(() => {
    const stores = makeStore()
    ontologyStore = stores.ontologyStore
    backingStore = stores.backingStore
  })

  describe('getType', () => {
    it('returns built-in node type without setup', async () => {
      const def = await ontologyStore.getType('built-in', 'entity.customer')
      expect(def).toBeDefined()
      expect(def!.type).toBe('entity.customer')
      expect(def!.status).toBe('active')
      expect(def!.label).toBeTruthy()
      expect(def!.description).toBeTruthy()
    })

    it('returns built-in edge type without setup', async () => {
      const def = await ontologyStore.getType('built-in', 'triggers')
      expect(def).toBeDefined()
      expect(def!.type).toBe('triggers')
      expect(def!.status).toBe('active')
    })

    it('returns undefined for nonexistent built-in type', async () => {
      const def = await ontologyStore.getType('built-in', 'nonexistent.type')
      expect(def).toBeUndefined()
    })

    it('returns type after upsert at brain scope', async () => {
      await ontologyStore.upsertType('brain', makeNodeDef())
      const def = await ontologyStore.getType('brain', 'entity.invoice')
      expect(def).toBeDefined()
      expect(def!.type).toBe('entity.invoice')
      expect(def!.label).toBe('Invoice')
    })

    it('returns undefined for nonexistent type at brain scope', async () => {
      const def = await ontologyStore.getType('brain', 'entity.missing')
      expect(def).toBeUndefined()
    })
  })

  describe('upsertType', () => {
    it('creates a new node type at brain scope', async () => {
      await ontologyStore.upsertType('brain', makeNodeDef())
      const def = await ontologyStore.getType('brain', 'entity.invoice')
      expect(def).toBeDefined()
      expect(def!.label).toBe('Invoice')
    })

    it('overwrites existing type with same ID', async () => {
      await ontologyStore.upsertType('brain', makeNodeDef({ label: 'V1' }))
      await ontologyStore.upsertType('brain', makeNodeDef({ label: 'V2' }))
      const def = await ontologyStore.getType('brain', 'entity.invoice')
      expect(def!.label).toBe('V2')
    })

    it('creates edge type correctly', async () => {
      await ontologyStore.upsertType('brain', makeEdgeDef())
      const def = await ontologyStore.getType('brain', 'manages')
      expect(def).toBeDefined()
      expect(def!.type).toBe('manages')
    })

    it('rejects upsert to built-in scope', async () => {
      await expect(
        ontologyStore.upsertType('built-in', makeNodeDef()),
      ).rejects.toThrow('cannot upsert to built-in scope')
    })

    it('rejects invalid type definition', async () => {
      const invalid: OntologyTypeDefinition = {
        type: 'entity.test',
        label: '',
        description: 'Missing label',
        createdAt: '2026-05-10T00:00:00.000Z',
        status: 'active',
      }
      await expect(ontologyStore.upsertType('brain', invalid)).rejects.toThrow()
    })
  })

  describe('deleteType', () => {
    it('removes node type from scope', async () => {
      await ontologyStore.upsertType('brain', makeNodeDef())
      await ontologyStore.deleteType('brain', 'entity.invoice')
      const def = await ontologyStore.getType('brain', 'entity.invoice')
      expect(def).toBeUndefined()
    })

    it('removes edge type from scope', async () => {
      await ontologyStore.upsertType('brain', makeEdgeDef())
      await ontologyStore.deleteType('brain', 'manages')
      const def = await ontologyStore.getType('brain', 'manages')
      expect(def).toBeUndefined()
    })

    it('throws for nonexistent type', async () => {
      await expect(
        ontologyStore.deleteType('brain', 'entity.nonexistent'),
      ).rejects.toThrow('not found')
    })

    it('rejects delete from built-in scope', async () => {
      await expect(
        ontologyStore.deleteType('built-in', 'entity.customer'),
      ).rejects.toThrow('cannot delete from built-in scope')
    })
  })

  describe('listTypes', () => {
    it('lists all 59 built-in types', async () => {
      const types = await ontologyStore.listTypes('built-in')
      expect(types).toHaveLength(59)
    })

    it('filters built-in types by prefix', async () => {
      const types = await ontologyStore.listTypes('built-in', { prefix: 'entity.' })
      expect(types).toHaveLength(6)
      for (const td of types) {
        expect(td.type.startsWith('entity.')).toBe(true)
      }
    })

    it('filters by status', async () => {
      await ontologyStore.upsertType('brain', makeNodeDef({ status: 'active' }))
      await ontologyStore.upsertType('brain', makeNodeDef({
        type: 'entity.proposed_one',
        label: 'Proposed',
        description: 'A proposed type',
        status: 'proposed',
      }))
      const proposed = await ontologyStore.listTypes('brain', { status: 'proposed' })
      expect(proposed).toHaveLength(1)
      expect(proposed[0]!.type).toBe('entity.proposed_one')
    })

    it('returns empty for scope with no types', async () => {
      const types = await ontologyStore.listTypes('brain')
      expect(types).toHaveLength(0)
    })
  })

  describe('getResolvedOntology', () => {
    it('returns all built-in types when store is empty', async () => {
      const resolved = await ontologyStore.getResolvedOntology('brain-1', 'proj-1', 'org-1')
      expect(resolved.nodeTypes).toHaveLength(30)
      expect(resolved.edgeTypes).toHaveLength(29)
      expect(resolved.businessCategories).toHaveLength(8)
    })

    it('brain scope overrides built-in', async () => {
      await ontologyStore.upsertType('brain', {
        type: 'entity.customer',
        label: 'Custom Customer',
        description: 'Brain-level override',
        createdAt: '2026-05-10T00:00:00.000Z',
        status: 'active',
      })
      const resolved = await ontologyStore.getResolvedOntology('brain-1', 'proj-1', 'org-1')
      const customer = resolved.nodeTypes.find((t) => t.type === 'entity.customer')
      expect(customer).toBeDefined()
      expect(customer!.label).toBe('Custom Customer')
      expect(customer!.scope).toBe('brain')
    })

    it('brain overrides project overrides org', async () => {
      await ontologyStore.upsertType('organisation', {
        type: 'entity.tenant',
        label: 'Org Tenant',
        description: 'Org level',
        createdAt: '2026-05-10T00:00:00.000Z',
        status: 'active',
      })
      await ontologyStore.upsertType('project', {
        type: 'entity.tenant',
        label: 'Project Tenant',
        description: 'Project level',
        createdAt: '2026-05-10T00:00:00.000Z',
        status: 'active',
      })
      await ontologyStore.upsertType('brain', {
        type: 'entity.tenant',
        label: 'Brain Tenant',
        description: 'Brain level',
        createdAt: '2026-05-10T00:00:00.000Z',
        status: 'active',
      })
      const resolved = await ontologyStore.getResolvedOntology('brain-1', 'proj-1', 'org-1')
      const tenant = resolved.nodeTypes.find((t) => t.type === 'entity.tenant')
      expect(tenant).toBeDefined()
      expect(tenant!.label).toBe('Brain Tenant')
      expect(tenant!.scope).toBe('brain')
    })

    it('project overrides org when no brain override', async () => {
      await ontologyStore.upsertType('organisation', {
        type: 'entity.account',
        label: 'Org Account',
        description: 'Org level',
        createdAt: '2026-05-10T00:00:00.000Z',
        status: 'active',
      })
      await ontologyStore.upsertType('project', {
        type: 'entity.account',
        label: 'Project Account',
        description: 'Project level',
        createdAt: '2026-05-10T00:00:00.000Z',
        status: 'active',
      })
      const resolved = await ontologyStore.getResolvedOntology('brain-1', 'proj-1', 'org-1')
      const account = resolved.nodeTypes.find((t) => t.type === 'entity.account')
      expect(account).toBeDefined()
      expect(account!.label).toBe('Project Account')
      expect(account!.scope).toBe('project')
    })

    it('excludes deprecated types from resolution', async () => {
      await ontologyStore.upsertType('brain', {
        type: 'entity.legacy',
        label: 'Legacy',
        description: 'Deprecated type',
        createdAt: '2026-05-10T00:00:00.000Z',
        status: 'deprecated',
      })
      const resolved = await ontologyStore.getResolvedOntology('brain-1', 'proj-1', 'org-1')
      const legacy = resolved.nodeTypes.find((t) => t.type === 'entity.legacy')
      expect(legacy).toBeUndefined()
    })

    it('includes custom business categories from all scopes', async () => {
      const stored = {
        customNodeTypes: [],
        customEdgeTypes: [],
        customBusinessCategories: ['logistics', 'healthcare'],
      }
      const data = Buffer.from(JSON.stringify(stored), 'utf-8')
      await backingStore.write(toPath('ontology/org/org-1/types.json'), data)

      const resolved = await ontologyStore.getResolvedOntology('brain-1', 'proj-1', 'org-1')
      expect(resolved.businessCategories).toContain('logistics')
      expect(resolved.businessCategories).toContain('healthcare')
      expect(resolved.businessCategories).toContain('customer')
    })

    it('caches resolved ontology', async () => {
      const resolved1 = await ontologyStore.getResolvedOntology('brain-1', 'proj-1', 'org-1')
      const resolved2 = await ontologyStore.getResolvedOntology('brain-1', 'proj-1', 'org-1')
      expect(resolved1).toBe(resolved2)
    })

    it('invalidates cache on upsert', async () => {
      const resolved1 = await ontologyStore.getResolvedOntology('brain-1', 'proj-1', 'org-1')
      await ontologyStore.upsertType('brain', makeNodeDef())
      const resolved2 = await ontologyStore.getResolvedOntology('brain-1', 'proj-1', 'org-1')
      expect(resolved1).not.toBe(resolved2)
    })

    it('invalidates cache on delete', async () => {
      await ontologyStore.upsertType('brain', makeNodeDef())
      const resolved1 = await ontologyStore.getResolvedOntology('brain-1', 'proj-1', 'org-1')
      await ontologyStore.deleteType('brain', 'entity.invoice')
      const resolved2 = await ontologyStore.getResolvedOntology('brain-1', 'proj-1', 'org-1')
      expect(resolved1).not.toBe(resolved2)
    })
  })
})
